libkbfs/block_journal.go

```diff
@@ -372,141 +224,50 @@ func (j *blockJournal) end() (journalOrdinal, error) {
 	return last + 1, nil
 }
 
-func (j *blockJournal) getRefEntry(
-	id BlockID, refNonce BlockRefNonce) (blockRefEntry, error) {
-	refs := j.refs[id]
-	if refs == nil {
-		return blockRefEntry{}, blockNonExistentError{id}
-	}
-
-	e, ok := refs[refNonce]
-	if !ok {
-		return blockRefEntry{}, blockNonExistentError{id}
-	}
-
-	return e, nil
-}
-
-func (j *blockJournal) putRefEntry(
-	id BlockID, refEntry blockRefEntry, ordinal journalOrdinal) error {
-	existingRefEntry, err := j.getRefEntry(
-		id, refEntry.context.GetRefNonce())
-	var exists bool
-	switch err.(type) {
-	case blockNonExistentError:
-		exists = false
-	case nil:
-		exists = true
-	default:
-		return err
-	}
-
-	if exists {
-		err = existingRefEntry.checkContext(refEntry.context)
-		if err != nil {
-			return err
-		}
-	}
-
-	if j.refs[id] == nil {
-		j.refs[id] = make(blockRefMap)
-	}
-
-	return j.refs[id].put(refEntry.context, refEntry.status, ordinal)
-}
-
-func (j *blockJournal) exists(id BlockID) error {
-	_, err := os.Stat(j.blockDataPath(id))
-	return err
+func (j *blockJournal) hasData(id BlockID) error {
+	return j.s.hasData(id)
 }
 
-func (j *blockJournal) getDataSize(id BlockID) (int64, error) {
-	fi, err := os.Stat(j.blockDataPath(id))
+func (j *blockJournal) remove(id BlockID) (int64, error) {
+	flushedBytes, err := j.s.getDataSize(id)
 	if err != nil {
 		return 0, err
 	}
-	return fi.Size(), nil
-}
-
-func (j *blockJournal) getData(id BlockID) (
-	[]byte, kbfscrypto.BlockCryptKeyServerHalf, error) {
-	data, err := ioutil.ReadFile(j.blockDataPath(id))
-	if os.IsNotExist(err) {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
-			blockNonExistentError{id}
-	} else if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
 
-	keyServerHalfPath := j.keyServerHalfPath(id)
-	buf, err := ioutil.ReadFile(keyServerHalfPath)
-	if os.IsNotExist(err) {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
-			blockNonExistentError{id}
-	} else if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	// Check integrity.
-
-	dataID, err := j.crypto.MakePermanentBlockID(data)
+	// TODO: we'll eventually need a sweeper to clean up entries
+	// left behind if we crash here.
+	err = j.s.remove(id)
 	if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	if id != dataID {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, fmt.Errorf(
-			"Block ID mismatch: expected %s, got %s", id, dataID)
+		return 0, err
 	}
 
-	var serverHalf kbfscrypto.BlockCryptKeyServerHalf
-	err = serverHalf.UnmarshalBinary(buf)
-	if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
+	if j.cachedUnflushedBytes >= 0 {
+		// If we somehow get out of sync and dip below zero,
```

Comment:
This is a nice property, but if there's any chance things get out of sync, it could just as easily always stay positive and never get fixed, right?  So hopefully we're not counting on this property.